modules/core/shared/src/main/scala/gem/util/Timestamp.scala

```diff
@@ -0,0 +1,127 @@
+  * resolution.  This allows Timestamps to roundtrip to/from the database
+  * where only microsecond resolution is supported.  In addition the min
+  * and max instants that are supported are determined by the postgres limit for
+  * timestamps.
+  *
+  * @param toInstant
+  */
+final class Timestamp private(val toInstant: Instant) extends AnyVal {
+
+  /** Gets the number of seconds from the Java epoch of 1970-01-01T00:00:00Z. */
+  def epochSecond: Long =
+    toInstant.getEpochSecond
+
+  /** Gets the number of microseconds after the start of the second returned
+    * by `epochSecond`.
+    */
+  def µs: Long =
+    toInstant.getNano / 1000L
+
+  /** Converts this instant to the number of milliseconds from the epoch of
+    * 1970-01-01T00:00:00Z.
+    */
+  def toEpochMilli: Long =
+    toInstant.toEpochMilli
+
+  /** Creates an updated instance of Timestamp by applying the given
+    * function to its wrapped Instant.  The computed Instant is truncated to
+    * microsecond precision and validated.
+    */
+  private def mod(f: Instant => Instant): Option[Timestamp] =
+    Timestamp.fromInstant(f(toInstant))
+
+  def plusMillis(millisToAdd: Long): Option[Timestamp] =
+    mod(_.plusMillis(millisToAdd))
+
+  def plusMicros(microsToAdd: Long): Option[Timestamp] =
+    mod(_.plusNanos(microsToAdd * 1000))
+
+  def plusSeconds(secondsToAdd: Long): Option[Timestamp] =
+    mod(_.plusSeconds(secondsToAdd))
+
+  override def toString: String =
+    toInstant.toString
+}
+
+object Timestamp {
+
+  /** Minimum time that can be stored in a postgres timestamp. */
+  val Min: Timestamp =
+    truncate(ZonedDateTime.of( -4712, 1, 1, 0, 0, 0, 0, UTC).toInstant)
+
+  /** Maximum time that can be stored in a postgres timestamp. */
+  val Max: Timestamp =
+    truncate(ZonedDateTime.of(294275, 12, 31, 23, 59, 59, 999999000, UTC).toInstant)
+
+  /** `Instant.EPOCH` transformed to `Timestamp`. */
+  val Epoch: Timestamp =
+    truncate(Instant.EPOCH)
+
+  /** Creates a Timestamp from the given Instant, assuring that the time
+    * value recorded has a round number of microseconds and that it is within
+    * the range supported by postgres.
+    *
+    * @group Constructors
+    */
+  def fromInstant(i: Instant): Option[Timestamp] = {
+    val iʹ = truncate(i)
+    if ((iʹ < Min) || (Max < iʹ)) None else Some(iʹ)
+  }
+
+  /** Creates a Timestamp from the given Instant if possible, throwing an
+    * exception if the time is out of the valid range `Min` <= time <= `Max`.
+    *
+    * @group Constructors
+    */
+  def unsafeFromInstant(i: Instant): Timestamp =
+    fromInstant(i).getOrElse(sys.error(s"$i out of Timestamp range"))
+
+  /** Creates a Timestamp from the given Instant, assuring that the time value
+    * recorded has a round number of microseconds. The caller must verify that
+    * the timestamp is in the valid range.
+    */
+  private def truncate(i: Instant): Timestamp =
+    new Timestamp(i.truncatedTo(MICROS))
```

Comment:
Can we inline this into `fromInstant` and then use `unsafeFromInstant` in the other cases here that use `truncate`? I would prefer to have the scale and limit validation together.